impl/src/metadata.rs

```diff
@@ -50,25 +97,132 @@ pub struct CargoWorkspaceFiles {
 /** A workspace metadata fetcher that uses the Cargo Metadata subcommand. */
 pub struct CargoMetadataFetcher {
   cargo_bin_path: PathBuf,
-  use_tempdir: bool,
+  registry_url: Url,
+  index_url: String,
 }
 
 impl CargoMetadataFetcher {
-  pub fn new<P: Into<PathBuf>>(cargo_bin_path: P, use_tempdir: bool) -> CargoMetadataFetcher {
+  pub fn new<P: Into<PathBuf>>(
+    cargo_bin_path: P,
+    registry_url: Url,
+    index_url: &str,
+  ) -> CargoMetadataFetcher {
     CargoMetadataFetcher {
       cargo_bin_path: cargo_bin_path.into(),
-      use_tempdir,
+      registry_url: registry_url.clone(),
+      index_url: String::from(index_url),
     }
   }
 
-  fn clone_crate_src(
-    &self,
-    registry_url: &str,
-    dir: &Path,
-    name: &str,
-    version: &str,
-  ) -> Result<PathBuf> {
-    let mut url = url::Url::parse(registry_url)?;
+  // Create a symlink file on unix systems
+  #[cfg(target_family = "unix")]
+  fn make_symlink(&self, src: &Path, dest: &Path) -> Result<()> {
+    std::os::unix::fs::symlink(src, dest)
+      .with_context(|| "Failed to create symlink for generating metadata")
+  }
+
+  // Create a symlink file on windows systems
+  #[cfg(target_family = "windows")]
+  fn make_symlink(&self, src: &Path, dest: &Path) -> Result<()> {
+    std::os::windows::fs::symlink_file(dest, src)
+      .with_context(|| "Failed to create symlink for generating metadata")
+  }
+
+  fn make_temp_workspace(&self, files: &CargoWorkspaceFiles) -> Result<(TempDir, PathBuf)> {
+    let temp_dir = TempDir::new()?;
+    assert!(files.toml_path.is_file());
+    assert!(files.lock_path_opt.as_ref().map_or(true, |p| p.is_file()));
+
+    // First gather metadata without downloading any dependencies so we can identify any path dependencies.
+    let no_deps_metadata = MetadataCommand::new()
+      .cargo_path(&self.cargo_bin_path)
+      .current_dir(files.toml_path.parent().unwrap()) // UNWRAP: Guarded by function assertion
+      .no_deps()
+      .exec()?;
+
+    // There should be a `Cargo.toml` file in the workspace root
+    fs::copy(
+      no_deps_metadata.workspace_root.join("Cargo.toml"),
+      temp_dir.as_ref().join("Cargo.toml"),
+    )?;
+
+    // Optionally copy over the lock file
+    if no_deps_metadata.workspace_root.join("Cargo.lock").exists() {
+      fs::copy(
+        no_deps_metadata.workspace_root.join("Cargo.lock"),
+        temp_dir.as_ref().join("Cargo.lock"),
+      )?;
+    }
+
+    // Copy over the Cargo.toml files of each workspace member
+    let re = Regex::new(r".+\(path\+file://(.+)\)")?;
+    for member in no_deps_metadata.workspace_members.iter() {
+      // Get a path to the workspace member directory
+      let path = {
+        let capture = match re.captures(&member.repr) {
+          Some(capture) => capture,
+          None => continue,
+        };
+
+        let re_match = match capture.get(1) {
+          Some(re_match) => re_match,
+          None => continue,
+        };
```

Comment:
Why not my version?

Reply:
There must be something wrong with the UI. Try opening the diff view. Do you see the full comment chain then?

Reproducing the comment:

This above can optionally be condensed.
```
let filename = {
  let filename_match =  re.captures(&member.repr).and_then(|c| c.get(1));
  if (filename_match.is_none()) {
    continue;
  }
  PathBuf::from(filename_match.unwrap()); // UNWRAP: Guarded above
};
```
(i don't really feel strongly about the forumulation of the unwrap -- more important is that you can use a map to avoid the second match)